Update android benchmark to alpha06 and parameterize tests

ZacSweers opened the pull request:
This revamps the design of the benchmark to be based on `times`. In short: 
* `times=0` means we're just measuring subscribe cost. No throughput of the observable stream, and it's disposed immediately un-timed.
* `times=n` where n is not 0 is full execution cost, with simple/complex/e2e variants. Tests have 1, 1000, and 1000000 versions.
* Filter out irrelevant variants, such as combinations where RDT is disabled but guarded delegates is enabled.

This does _not_ update the JMH tests yet, but we should do that too and add it to `benchmark.md`.

Updated results: see `benchmark.md` changes.

ShaishavGandhi commented on android-benchmark/benchmark/build.gradle:
```diff
@@ -37,12 +37,19 @@ android {
-  implementation "androidx.test:rules:1.1.1"
-  implementation "androidx.test.ext:junit:1.1.0"
-  implementation "org.jetbrains.kotlin:kotlin-stdlib:1.3.31"
+  implementation "androidx.benchmark:benchmark-common:${deps.versions.androidBenchmark}"
```
Let's pull these dependencies out?

ZacSweers replied:
I don't follow, the version is already shared

ShaishavGandhi replied:
I meant the maven coordinates themselves. Or do you usually wait for the dependency to show up in a second place before extracting them into `dependencies.gradle`?

ZacSweers replied:
Yeah for gradle plugin dependencies like these that are only defined once, I'm not too worried about moving the whole coordinate into dependencies.gradle

ShaishavGandhi commented on android-benchmark/benchmark/src/androidTest/kotlin/com/uber/rxdogtag/androidbenchmark/RxDogTagAndroidPerf.kt:
```diff
@@ -39,214 +43,62 @@ import java.util.concurrent.CountDownLatch
+    @JvmStatic
+    @Parameters(name = "enabled={0},times={1}")
+    fun data() = listOf(
+        arrayOf(true, 0),
```
Correct me if I'm wrong, but wouldn't we be generating lot of tests that we don't care about in this parameterized model? 

For example, I'll have a boolean for whether guardedDelegateCall is run. 

That will generate a redundant set permutation of RxDogTag being disabled and guardedDelegate being enabled. 

This will only increase as we add more parameters.

ZacSweers replied:
When that time comes, it's easy to add an annotation + rule that ignores tests for certain configurations. Example: https://github.com/square/kotlinpoet/blob/master/kotlinpoet/src/test/java/com/squareup/kotlinpoet/KmSpecsTest.kt#L84

ShaishavGandhi replied:
Cool, that's neat.

ShaishavGandhi commented:
Seems like the new benchmark is off?

```
1000 items (flowable)
RxDogTagAndroidPerf.flowable_complex[enabled=true,times=1,000]               51,131ns    0.051ms
RxDogTagAndroidPerf.flowable_e2e[enabled=true,times=1,000]                  163,333ns    0.163ms  219.44%
RxDogTagAndroidPerf.flowable_simple[enabled=true,times=1,000]             2,953,542ns    2.954ms  5676.42%
```

This is basically comparing a complex Flowable with RxDogTag enabled with a flowable e2e with RxDogTag enabled?

Wouldn't we want to compare a complex Flowable with RxDogTag enabled with a complex Flowable with RxDogTag disabled?

ZacSweers commented:
Hmm yeah. Doubt it's the benchmark, but rather maybe the parser is dropping data. Will look

ShaishavGandhi commented:
Can you post the new results with the DataParser bug fixed?

ZacSweers commented:
Will do, haven't fixed the bug yet though.

On Tue, Aug 20, 2019 at 12:58 AM Shaishav Gandhi <notifications@github.com>
wrote:

> Can you post the new results with the DataParser bug fixed?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/uber/RxDogTag/pull/41?email_source=notifications&email_token=AAKMJPRMCZNZ66F5IN4NITTQFN2W3A5CNFSM4ILQQFEKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4VBXLI#issuecomment-522853293>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAKMJPVM4JPPOVEGUUBBR6DQFN2W3ANCNFSM4ILQQFEA>
> .
>

ShaishavGandhi commented on docs/benchmark.md:
```diff
@@ -12,86 +12,202 @@ It's important to look at the units than at the percentages. In general, RxDogTa
+| RxDogTagAndroidPerf.observable_simple[enabled=false,times=1,guardedDelegateEnabled=false]          |         577ns |       0.001ms |
+| RxDogTagAndroidPerf.observable_simple[enabled=true,times=1,guardedDelegateEnabled=false]           |      13,903ns |       0.014ms | 2309.53% |
+| RxDogTagAndroidPerf.observable_simple[enabled=true,times=1,guardedDelegateEnabled=true]            |      14,303ns |       0.014ms | 2378.86% |
+| RxDogTagAndroidPerf.observable_simple[enabled=false,times=1,000,guardedDelegateEnabled=false]      |      23,987ns |       0.024ms | 4057.19% |
```
The heading says: `1 item (Observable)` but the benchmark has ones where we're sending 1000 and 1_000_000. We should filter these out

ShaishavGandhi commented on docs/benchmark.md:
```diff
@@ -12,86 +12,202 @@ It's important to look at the units than at the percentages. In general, RxDogTa
+| RxDogTagAndroidPerf.flowable_e2e[enabled=true,times=1,000,000,guardedDelegateEnabled=false]        | 654,682,512ns |      654.682ms | 119905121.98% |
+| RxDogTagAndroidPerf.flowable_complex[enabled=true,times=1,000,000,guardedDelegateEnabled=true]     | 755,996,898ns |      755.997ms | 138460870.33% |
+
+#### 1_000 items (Observable)
```
Same here as we're showing results for 1_000_000 items (which has a separate section anyway)

ShaishavGandhi commented on docs/benchmark.md:
```diff
@@ -12,86 +12,202 @@ It's important to look at the units than at the percentages. In general, RxDogTa
+| RxDogTagAndroidPerf.observable_e2e[enabled=true,times=1,guardedDelegateEnabled=false]              |     119,878ns |     0.120ms |   56.98% |
+| RxDogTagAndroidPerf.observable_e2e[enabled=true,times=1,guardedDelegateEnabled=true]               |     126,007ns |     0.126ms |   65.01% |
+| RxDogTagAndroidPerf.observable_e2e[enabled=true,times=0,guardedDelegateEnabled=false]              |     135,660ns |     0.136ms |   77.65% |
+| RxDogTagAndroidPerf.observable_e2e[enabled=false,times=1,000,guardedDelegateEnabled=false]         |     320,312ns |     0.320ms |  319.46% |
```
Is it possible to break the E2E cost by number of items? For example it's not valuable to compare e2e cost of 1000 item Observable with RxDogTag disabled with 1 item Observable with RxDogTag enabled. Giving comparisons for same number of items will be more useful since the % change will make more sense there. 

Right now you have to do a lot of cognitive calculations to know what the real e2e cost of RxDogTag is since you subtract the 1000 Observable(with RxDogTag) with 1000 Observable(without RxDogTag)

ZacSweers commented:
Updated the PR with some changes:

- Removed e2e entirely as the `complex` type basically matches it now
- Added more subgroups to throughput instead, dynamically calculated
- Removed percent delta because it really encourages readers to focus on the wrong thing, when the main thing to look at is millisecond delta

ShaishavGandhi commented on docs/benchmark.md:
```diff
@@ -12,86 +12,119 @@ It's important to look at the units than at the percentages. In general, RxDogTa
 
 Running the benchmark on a [Pixel 3](https://store.google.com/product/pixel_3_specs) yields the following results:
```
Can't comment there but can you remove the note about percentages?

ZacSweers replied:
Done in dde7232

ShaishavGandhi commented on docs/benchmark.md:
```diff
@@ -12,86 +12,119 @@ It's important to look at the units than at the percentages. In general, RxDogTa
+#### Simple: 1 item (Observable)
+| Benchmark | Time (ns) | Time (ms) |
+|----------|------------|-----------|
+| RxDogTagAndroidPerf.observable_simple[enabled=false,times=1,guardedDelegateEnabled=false]          |         532ns |   0.001ms |
```
Minor nit: Can we shorten the name of the benchmark? Not sure how much it makes sense to have the `RxDogTagAndroidPerf.observable_simple` part anymore. Only seems to add noise.

ZacSweers replied:
Just did that in 61cd7eb!

![image](https://user-images.githubusercontent.com/1361086/65479390-2f643f80-de5b-11e9-83bc-de49cba481d9.png)
